src/transaction.rs

```diff
@@ -352,7 +345,7 @@ impl Savepoint<'_> {
         }
         match self.drop_behavior() {
             DropBehavior::Commit => self.commit_().or_else(|_| self.rollback()),
-            DropBehavior::Rollback => self.rollback(),
+            DropBehavior::Rollback => self.rollback().and_then(|_| self.commit_()),
```

Comment:
To be checked.
Related to https://github.com/rusqlite/rusqlite/issues/1326#issuecomment-1553157537 ?

Reply:
Yes, related to the comment.  Without a commit call, the savepoint is left on the stack, so dropping a savepoint in rollback mode without this change will leave it on the savepoint stack.  It's different from non-savepoint transactions, where the ROLLBACK also releases it.

In practice, this hasn't caused much of an issue, because the savepoint is removed when rolling back or committing a savepoint (or transaction) higher in the stack, so creating a savepoint and then not interacting with it at all in the future is usually functionally the same as not doing anything at all.  A spot where this could really make a difference is if you create a savepoint that's not in a transaction, then the drop of the savepoint will leave you in that transaction and not in autocommit mode like you might expect:

```rust
let mut conn = Connection::open(db_path)?;
assert!(conn.is_autocommit());
{
    let savepoint = conn.savepoint()?;
}
assert!(conn.is_autocommit());
```

That last assert fails.

As a result, this fails with QueryReturnedNoRows:

```rust
    {
        let mut conn = Connection::open(&db_path)?;
        conn.execute("CREATE TABLE sample (id INTEGER PRIMARY KEY)", ())?;
        {
            let savepoint = conn.savepoint()?;
        }
        conn.execute("INSERT INTO sample (id) VALUES (1)", ())?;
    }
    {
        let mut conn = Connection::open(&db_path)?;
        let id: i64 = conn.query_row("SELECT id FROM sample", (), |row| row.get(0))?;
    }
```

Because the savepoint wasn't released.

I assume this issue is the reason the savepoint depth was thought to be necessary, because with the depth, you can at least always commit or roll back the right savepoint, while ignoring the dangling ones higher up.  This works fine until you are not interacting with a higher transaction.

Reply:
1. No, with the depth, you still need to release the savepoint.  Notice https://sqlite.org/lang_savepoint.html

> The [ROLLBACK](https://sqlite.org/lang_transaction.html) command without a TO clause rolls backs all transactions and leaves the transaction stack empty.
> 
> The ROLLBACK command with a TO clause rolls back transactions going backwards in time back to the most recent SAVEPOINT with a matching name. The SAVEPOINT with the matching name remains on the transaction stack, but all database changes that occurred after that SAVEPOINT was created are rolled back. If the savepoint-name in a ROLLBACK TO command does not match any SAVEPOINT on the stack, then the ROLLBACK command fails with an error and leaves the state of the database unchanged.

To note: "The SAVEPOINT with the matching name remains on the transaction stack".  With the depth or without, there is a bug here.  My example code has the wrong behavior on the latest release version.

> The RELEASE command starts with the most recent addition to the transaction stack and releases savepoints backwards in time until it releases a savepoint with a matching savepoint-name. Prior savepoints, even savepoints with matching savepoint-names, are unchanged. If the RELEASE command causes the transaction stack to become empty (if the RELEASE command releases the outermost transaction from the stack) then the transaction commits.

2.  You don't actually need to remove the depth, but it's not buying any extra functionality or convenience.  It's extra stuff done in the code which partially masks the "release" bug.

Reply:
> 1. No, with the depth, you still need to release the savepoint.  Notice https://sqlite.org/lang_savepoint.html

So could you please provide a minimal unit test (that fails without `self.rollback().and_then(|_| self.commit_())` in current master or release branch) ?

Reply:
Yes, absolutely.  I should have done that in the first place.  It's added now.

Reply:
This is unexpected to me:
```sql
sqlite> CREATE TABLE t AS SELECT 1 as n;
sqlite> SAVEPOINT s; -- first savepoint
sqlite> UPDATE t SET n=2;
sqlite> SAVEPOINT s; -- second savepoint
sqlite> UPDATE t SET n=3;
sqlite> ROLLBACK TO s; -- second savepoint rollbacked
sqlite> RELEASE s; -- expected first savepoint to be released but seems to be the second one !!!
sqlite> RELEASE s; -- expected failure but first savepoint released
```

Reply:
Yes, I agree. It was also totally unexpected to me, and caused a lot of headaches for me when using savepoints heavily (not with rusqlite; it's a C++ codebase) until I discovered what was actually happening.  It's a strange and specific difference in semantics between `SAVEPOINT` and `TRANSACTION`.

I think the idea is that you could roll back the savepoint and then start it again from the beginning without having to re-create it.  I would have preferred consistency with other transaction operations, personally.